kolibri/core/assets/src/conditionalPromise.js

```diff
@@ -0,0 +1,51 @@
+class ConditionalPromise {
+  /**
+   * Create a conditional promise - like a promise, but cancelable!
+   */
+  constructor(...args) {
+    if ([...args].length) {
+      this._promise = new Promise(...args);
+    }
+  }
+
+  catch(...args) {
+    this._promise.catch(...args);
+    return this;
+  }
+
+  then(...args) {
+    this._promise.then(...args);
+    return this;
+  }
+
+  only(cancelCheck, resolve, reject) {
```

Comment:
I made it up myself. I did start looking at other suggestions for how this would be done (i.e. being able to call cancel on a promise and cancel all its handlers), but this seemed to be the lowest friction solution, that didn't require getting into the innards of the Promises.